Sqlite-command: Add/bdd tests

kozer opened the pull request:
This PR adds BDD ( behat ) tests for sqlite.
Fixes https://github.com/Automattic/wp-cli-sqlite-command/issues/3
### Testing
N/A

kozer committed:
Sqlite: Add import tests

kozer committed:
Sqlite: Add export tests

kozer committed:
Update: run tests on push

kozer committed:
Update: Fix lint errors

kozer committed:
Update: Fix lint errors

kozer committed:
Update: Fix lint errors

kozer committed:
Update: Fix lint errors

kozer committed:
Update: Fix lint errors

kozer committed:
Update: Fix lint errors

kozer committed:
Update: workflow to use php > 7.4

kozer committed:
Update: workflow to use php > 7.4

kozer committed:
Update: Force sqlite to tests

kozer committed:
Update: fix warning

kozer committed:
Update: testing keeping only sqlite results

kozer committed:
Update: Remove unecessary part

kozer committed:
Update: Run tests only on sqlite dbtype

kozer commented on src/SQLite_Command.php:
```diff
@@ -76,10 +76,14 @@ public function import( $args, $assoc_args ) {
 	 *  Success: Exported to 'wordpress_dbase.sql'.
 	 *
-	 * @when before_wp_load
+	 * @when after_wp_load
```
@jeroenpf, this change has been done, as in tests, `WP_SQLite_Translator` is instantiated as part of import and export. When the tests are running the `WP_SQLite_Translator`, try to call `do_action` which, in `before_wp_load` is undefined. That said, is it ok to call import and export after wp-load,  as above? Does this create a problem?
Here is part of the output:
```          thrown in /tmp/wp-cli-test-run--66ba10bc05cf53.68302961/wp-content/mu-plugins/sqlite-database-integration/wp-includes/sqlite/class-wp-sqlite-translator.php on line 3992

        Fatal error: Uncaught Error: Call to undefined function do_action() in /tmp/wp-cli-test-run--66ba10bc05cf53.68302961/wp-content/mu-plugins/sqlite-database-integration/wp-includes/sqlite/class-wp-sqlite-translator.php on line 3992

        Error: Call to undefined function do_action() in /tmp/wp-cli-test-run--66ba10bc05cf53.68302961/wp-content/mu-plugins/sqlite-database-integration/wp-includes/sqlite/class-wp-sqlite-translator.php```
Same goes with `import` as well.

jeroenpf replied:
This will not work. If you try to import into a corrupt/empty/incomplete database, the command will fail if you load wp first. Same goes for export, it should work regardless of the state of the Wordpress installation.

I also think you might have an old version of the sqlite integration plugin because do_action and apply_filters instances are only executed conditionally if they exist. Can you double check with which version of the sqlite integration plugin you are testing?

jeroenpf replied:
You might need to update wp-cli-tests repo. If you have an old version it will use a hardcoded version of sqlite integration plugin that is very outdated.

kozer replied:
Thanks for that @jeroenpf ! 
I manage to revert that by using the latest version.
Some notes:
1. The latest changes are not in a version, and I uses `dev-main` to make this work.
2. Updated the tests, as there were changes by using the latest sqlite plugin.
3. Reverted the `afte9r_wp_load` to `before_wp_load`.

kozer committed:
Update: fix tests

jeroenpf commented on src/SQLite_Command.php:
```diff
@@ -79,7 +79,11 @@ public function import( $args, $assoc_args ) {
+		$is_porcelain = isset( $assoc_args['porcelain'] );
+
+		if ( ! $is_porcelain ) {
+			WP_CLI::success( 'Exporting database...' );
```
Thanks for adding the porcelain check here. I actually just realize now that i left this output when i was testing but we should probably remove it.

kozer commented:
@jeroenpf , I need some assistant here. Although the tests are marked to run only on `slite` "mode", it seems that it still fails the tests.
Here is what I get when I run tests locally, on "MySQL mode":
```
No scenarios
No steps
0m0.19s (10.96Mb)
Script rerun-behat-tests handling the behat-rerun event returned with error code 1
```
Do you have any idea on how we surpass this? Thanks!

jeroenpf commented:
I think this might be because there are no tests at all for MySQL? the output you shared shows that no scenarios and no steps were executed, indicating that the exit code 1 is caused by there not being any tests at all. What if you create a simple scenario that runs for both MySQL and SQLite? Do you still get an error in that case?

kozer committed:
Update: Add a test that always passes to make jobs to suceed

kozer committed:
Update: Fix composer errors

kozer commented:
> I think this might be because there are no tests at all for MySQL? the output you shared shows that no scenarios and no steps were executed, indicating that the exit code 1 is caused by there not being any tests at all. What if you create a simple scenario that runs for both MySQL and SQLite? Do you still get an error in that case?

@jeroenpf , I updated the code, by adding one more test that pass each time.
I have also updated `composer.json`, which I want you to take a look, as I make the install to pass again,but I'm not sure if this is the best way to go.
Thanks!
